jupyter_server/services/contents/filemanager.py

```diff
@@ -55,6 +56,31 @@ def _validate_root_dir(self, proposal):
             raise TraitError("%r is not a directory" % value)
         return value
 
+    @default("preferred_dir")
+    def _default_preferred_dir(self):
+        try:
+            value = self.parent.preferred_dir
+        except AttributeError:
+            pass
+        else:
+            if value is not None:
+                warnings.warn(
+                    "ServerApp.preferred_dir config is deprecated in jupyter-server 2.0. Use ContentsManager.preferred_dir with a relative path instead",
+                    DeprecationWarning,
+                    stacklevel=3,
+                )
+                if not (value + os.path.sep).startswith(self.root_dir):
+                    raise TraitError("%s is outside root contents directory" % value)
```

Comment:
I'm happy to tweak the stacklevel. I agree that we should probably catch and treat TraitErrors differently to the "catch-all" part. I think your other concern will also be addressed if we switch to [`pathlib.PurePath.relative_to`](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.relative_to), since that will raise for any non-sub paths.

Reply:
> If you look at the Exception object, I'm sure you'll find it to be an instance of TraitError trying to convey that preferred_dir is outside the scope of root_dir. Perhaps it should just get re-raised when rectification still produces an invalid directory?

If that was the case, then it would use "/" as preferred dir, and you wouldn't see any errors in lab.

Reply:
I added `as ex` to the Exception block, here's the view from the debugger...
![image](https://user-images.githubusercontent.com/22599560/195992309-9cf6b00e-a8e3-4c67-962f-6cefdbf3bad2.png)